tutorial/lesson_18_parallel_reductions.cpp

```diff
@@ -0,0 +1,219 @@
+// Halide tutorial lesson 17: Factoring a reduction using rfactor
+
+// This lesson demonstrates basic usage of Halide as a JIT compiler for imaging.
+
+// On linux, you can compile and run it like so:
+// g++ lesson_18*.cpp -g -I ../include -L ../bin -lHalide -lpthread -ldl -o lesson_18 -std=c++11
+// LD_LIBRARY_PATH=../bin ./lesson_18
+
+// On os x:
+// g++ lesson_18*.cpp -g -I ../include -L ../bin -lHalide -o lesson_18 -std=c++11
+// DYLD_LIBRARY_PATH=../bin ./lesson_18
+
+// If you have the entire Halide source tree, you can also build it by
+// running:
+//    make tutorial_lesson_18_parallel_reductions
+// in a shell with the current directory at the top of the halide
+// source tree.
+
+// The only Halide header file you need is Halide.h. It includes all of Halide.
+#include "Halide.h"
+
+// We'll also include stdio for printf.
+#include <stdio.h>
+
+using namespace Halide;
+
+// Support code for loading pngs.
+#include "halide_image_io.h"
+using namespace Halide::Tools;
+
+int main(int argc, char **argv) {
+    // Declare some Vars to use below.
+    Var x("x"), y("y"), u("u");
+
+    // Load a grayscale image to use as an input.
+    Buffer<uint8_t> input = load_image("images/gray.png");
+
+    {
+        // As mentioned previously in lesson 9, parallelizing variables that
+        // are part of the reduction domain is tricky, since there may be data
+        // dependencies across those variables.
+        //
+        // Consider the histogram example in lesson 9:
+        //
+        // Func histogram("histogram");
+        // Var x("x");
+        // histogram(x) = 0;
+        // RDom r(0, input.width(), 0, input.height());
+        // histogram(input(r.x, r.y)) += 1;
```

Comment:
Perhaps these code snippets, even though they're old examples, should be real code, not comments? This would cause the compiler to check them, making sure they don't rot, and would let them be syntax highlighted for better readability. (This would require splitting this anonymous block into three blocks, for the redefinitions of `Func histogram`, etc.)